### PR TITLE
Give more control over which variables mutate() should keep

### PR DESCRIPTION
**Pull request by hadley**
This is an extremely quick and hacky implementation of `morph()`, as described in #3721.

### CODE REVIEW

**Comment by hadley** [comment]
I think I'll make this a clearly marked experiment argument of `mutate()` that only applies to data frames. If it seems successful (and isn't too hard to implement later) we can broaden it, but for now it seems reasonable to keep it as tightly scoped as possible.

I'll also add tests that use `across()` and `cur_data()` so we don't accidentally break in the future.

**Comment by hadley** [comment]
Actually I think `cur_data()` will have to trigger variable usage, at least until we figure out how to make subsetting lazier.

File: R/data-mask.R [hunk 1]
```diff
@@ -24,12 +24,23 @@ DataMask <- R6Class("DataMask",
...
       }
 
-      binding_fn <- function(index, chunks = resolve_chunks(index)){
-        # chunks is a promise of the list of all chunks for the column
-        # at this index, so resolve_chunks() is only called when
-        # the active binding is touched
-        function() .subset2(chunks, private$current_group)
+      if (track_usage) {
+        private$used <- rep(FALSE, ncol(data))
+        binding_fn <- function(index, chunks = resolve_chunks(index)) {
```
Comment (krlmlr) [inline R/data-mask.R]: Is it safer to `force(index)` here? I remember https://github.com/tidyverse/ggplot2/pull/2807 caused by missing `force()` when returning a closure, but I haven't found an instance where this implementation breaks.

File: R/data-mask.R [hunk 2]
```diff
@@ -24,12 +24,23 @@ DataMask <- R6Class("DataMask",
...
+          function() {
+            private$used[[index]] <- TRUE
+            .subset2(chunks, private$current_group)
+          }
+        }
+      } else {
+        binding_fn <- function(index, chunks = resolve_chunks(index)){
```
Comment (krlmlr) [inline R/data-mask.R]: Same here?

**Comment by hadley** [comment]
@mine-cetinkaya-rundel suggested another interesting variation — show the only new variables and the existing variables used to compute them. Maybe `remove = "unused"`

File: R/data-mask.R (hunk 1 above)
Comment (hadley) [reply R/data-mask.R]: Given that it works, I think we can leave as is

**Pull request merged**